packages/sdk/src/react/_internal/transaction-machine/execute-transaction.ts

```diff
@@ -510,94 +596,169 @@ export class TransactionMachine {
 		await this.openPaymentModalWithPromise(paymentModalProps);
 	}
 
-	private async executeStep({
-		step,
+	private async execute({
 		props,
 	}: {
-		step: Step;
-		props: TransactionInput['props'];
-	}) {
-		debug('Executing step', { stepId: step.id });
-		if (!step.to && !step.signature) {
+		props: TransactionInput;
+	}): Promise<Result<TransactionState>> {
+		const { steps } = this.transactionState;
+
+		debug('Executing transaction', { props, steps });
+
+		if (!steps.steps) {
+			throw new Error('No steps found');
+		}
+
+		const executionStep = steps.steps[0];
+		debug('Executing step', { stepId: executionStep.id });
+
+		if (!executionStep.to && !executionStep.signature) {
 			throw new Error('Invalid step data');
 		}
 
-		try {
-			await this.switchChain();
-			if (step.id === StepType.buy) {
-				await this.executeBuyStep({ step, props: props as BuyInput });
-			} else if (step.signature) {
-				await this.executeSignature(step);
-			} else if (step.id === StepType.tokenApproval) {
-				//TODO: Add some sort ofs callback heres
-				const hash = await this.executeTransaction(step);
-				return { hash }
+		this.transition((prevState) => ({
+			...prevState,
+			transaction: {
+				...prevState.transaction,
+				ready: true,
+				executing: true,
+				executed: false,
+			},
+		}));
+
+		return this.executeOperation(async () => {
+			if (executionStep.id === StepType.buy) {
+				await this.executeBuyStep({
+					step: executionStep,
+					props: props.props as BuyInput,
+				});
+			} else if (executionStep.signature) {
+				await this.executeSignature(executionStep);
 			} else {
-				const hash = await this.executeTransaction(step);
-				this.config.onSuccess?.(hash);
-				return { hash };
+				await this.executeTransaction({
+					step: executionStep,
+					isTokenApproval: false,
+				});
 			}
-		} catch (error) {
-			this.config.onError?.(error as Error);
-			throw error;
-		}
+
+			return {
+				...this.transactionState,
+				transaction: {
+					...this.transactionState.transaction,
+					ready: false,
+					executing: false,
+					executed: true,
+				},
+			};
+		});
 	}
 
-	async getTransactionSteps(
-		props: TransactionInput['props'],
-	): Promise<TransactionSteps> {
+	private async approve({
+		step,
+	}: {
+		step: Step;
+	}): Promise<Result<TransactionState>> {
+		return this.executeOperation(async () => {
+			await this.transition((prevState) => ({
+				...prevState,
+				approval: {
+					...prevState.approval,
+					needed: true,
+					processing: true,
+					processed: false,
+				},
+			}));
+
+			debug('Executing step', { stepId: step.id });
+
+			if (!step.to && !step.signature) {
```

Comment:
This looks wrong, I dont think we ever will get an off-chain signature as part of an approval request?